Persist S3 client through Bucket __init__

jimas14 opened the pull request:
Our team was tasked to upgrade Thumbor (from 6.7.0) to be able to preserve IPTC metadata. This requires a custom engine https://github.com/scorphus/thumbor-wand-engine, which requires python 3, which requires Thumbor 7.0.0a5. After putting all the pieces together, including the code from @amanagr's PR https://github.com/thumbor-community/aws/pull/147 to accommodate async Thumbor 7, things worked, but we noticed a ton of asyncio errors and a minimum response time of > 3000ms on load testing. 

We utilize S3 loader & result storage for our requests, and we noticed that the S3 client creation/Bucket initialization was being called upwards of 5 times for each request.

We took a pass at trying to improve this by only creating the S3 client on Bucket's first __init__ call, and reusing that client for every subsequent request. These changes alone took the minimum response time down to < 200ms (like our 6.7.0 implementation did), and performed very well in load tests (15 containers, numprocs=2, 1000 random requests by 10 users -- same parameters as first load test).

I'm very interested to see what you all think. There was a couple other small refactors added as well. Credits to @cselvaraj for the idea & implementation.

cc @amanagr @kkopachev

kkopachev commented on tc_aws/aws/bucket.py:
```diff
@@ -121,7 +125,9 @@ async def delete(self, path):
 
-    def _clean_key(self, path):
+    @staticmethod
+    @lru_cache()
```
what's that for?

cselvaraj replied:
A very minor performance change.  Nothing important really.

kkopachev commented on tc_aws/aws/bucket.py:
```diff
@@ -34,23 +42,20 @@ def __init__(self, bucket, region, endpoint, max_retry=None):
-            endpoint_url=endpoint,
-            config=config
-        )
+        if self._client is None:
```
so, from what I understand, this is the only required change here. Other changes are rather cosmetic.

cselvaraj replied:
Yes, this is the operative change.  The rest are PEP8 changes and very minor performance changes.

kkopachev reviewed (approved):
This looks good. I'd only keep `Bucket.__init__` changes as most relevant.

kkopachev commented:
Ultimate solution to this trouble would be to modify Thumbor so it is able to call preinit methods on storage and other modules. And cleanup right before server shutdown.

jimas14 commented:
@kkopachev Thanks so much for the review. I just updated the PR if you could take another look. I updated the PR description with our updated findings.

kkopachev commented:
This is awesome @jimas14! Great catch on HeadObject operation!
My only concern here is that we're pinning aibotocore to an older version.

peterrus commented:
Potentially relevant: https://github.com/thumbor-community/aws/pull/147#issuecomment-1029186124
